src/testsets.jl

```diff
@@ -2,42 +2,97 @@ import Test: Test, finish, record
 
 using Test
 using Test: AbstractTestSet
+using ReTestItems:
+    JUnitTestSuites, JUnitTestSuite, JUnitTestCase, write_junit_file, junit_record!
+import ReTestItems
 
-# Wraps a DefaultTestSet and adds a list of Tasks for concurrent tests
-mutable struct ConcurrentTestSet <: Test.AbstractTestSet
+mutable struct RAITestSet <: Test.AbstractTestSet
     dts::Test.DefaultTestSet
-    tests::Vector{Task}
+    report::Bool
+    distributed::Bool
+    distributed_tests::Vector{Task}
+    junit::Union{JUnitTestSuites, JUnitTestSuite}
+    name_dict::Dict{String, Int}
+
+    function RAITestSet(dts, report, distributed, name_dict)
+        desc = dts.description
+        if Test.get_testset_depth() == 0
+            junit = JUnitTestSuites(desc)
+        else
+            junit = JUnitTestSuite(desc)
+        end
+        return new(dts, report, distributed, [], junit, name_dict)
+    end
+end
+
+function RAITestSet(desc; report::Option{Bool}=nothing, distributed::Option{Bool}=nothing)
+    dts = Test.DefaultTestSet(desc)
+    is_nested = Test.get_testset_depth() > 0
+    default_report = false
+    default_distributed = true
+    default_name_dict = Dict{String, Int}()
+
+    # Pass on the parent RAITestSet's options if nested
+    if is_nested
+        parent = Test.get_testset()
+        if parent isa RAITestSet
+            default_report = parent.report
+            default_distributed = parent.distributed
+            default_name_dict = parent.name_dict
+        end
+    end
+
+    return RAITestSet(
+        dts,
+        something(report, default_report),
+        something(distributed, default_distributed),
+        default_name_dict,
+    )
+end
+
+is_distributed(ts::RAITestSet) = ts.distributed
+is_distributed(ts::Test.AbstractTestSet) = false
 
-    ConcurrentTestSet(desc) = new(Test.DefaultTestSet(desc), [])
+is_reportable(ts::RAITestSet) = ts.report
+is_reportable(ts::Test.AbstractTestSet) = false
+
+function distribute_test(f, ts::RAITestSet)
+    ref = Threads.@spawn f()
+    return push!(ts.distributed_tests, ref)
 end
 
-record(ts::ConcurrentTestSet, child::AbstractTestSet) = record(ts.dts, child)
-record(ts::Test.DefaultTestSet, child::ConcurrentTestSet) = record(ts, child.dts)
-record(ts::ConcurrentTestSet, res::Test.Result) = record(ts.dts, res)
+function record(ts::RAITestSet, child::RAITestSet)
+    junit_record!(ts.junit, child.junit)
+    return record(ts.dts, child.dts)
+end
+record(ts::RAITestSet, child::AbstractTestSet) = record(ts.dts, child)
+record(ts::Test.DefaultTestSet, child::RAITestSet) = record(ts, child.dts)
+record(ts::RAITestSet, res::Test.Result) = record(ts.dts, res)
 
 # Record any results directly stored and fetch results from any listed concurrent tests
 # If this is the parent then show results
-function finish(ts::ConcurrentTestSet)
-    for t in ts.tests
-        record(ts.dts, fetch(t))
+function finish(ts::RAITestSet)
+    ts.dts.time_end = time()
```

Comment:
maybe add a comment why we're doing this here?

presumably the final `finish(ts.dts)` call at the bottom of the function is going to do this (and over-write this value)

(i'm guessing the answer is so that we have the `time_end` set before calling `write_junit_file`, so we can get timings?)

But why not just call `finish(ts.dts)` before the `is_reportable` block? I guess because `finish` also does printing of the results and maybe we want to printing of results _before_ the JUnit writing?

Reply:
As you guessed, we need the time end before the writing the junit file. If there are any failures, calling `finish(ts.dts)` throws if it's the root testset, so then we don't get any report. So it's a bit of a dance of getting the data we need to report on and doing reporting before we can call `finish`.

Reply:
Also, the way we asynchronously run tests means we lose the connection between a `TestRelTestSet` and its parent, so if we call finish on those, the default test set thinks it's the root and prints the results and throws an error as such. So we don't always finish those test sets.